acvm-repo/acvm/tests/solver.rs

```diff
@@ -722,3 +726,223 @@ fn memory_operations() {
 
     assert_eq!(witness_map[&Witness(8)], FieldElement::from(6u128));
 }
+
+// Solve the given BlackBoxFuncCall with witnesses: 1, 2 as x, y, resp.
+#[cfg(test)]
+fn solve_blackbox_func_call(
+    blackbox_func_call: impl Fn(
+        Option<FieldElement>,
+        Option<FieldElement>,
+    ) -> BlackBoxFuncCall<FieldElement>,
+    x: (FieldElement, bool), // if false, use a Witness
+    y: (FieldElement, bool), // if false, use a Witness
+) -> FieldElement {
+    let (x, x_constant) = x;
+    let (y, y_constant) = y;
+
+    let initial_witness = WitnessMap::from(BTreeMap::from_iter([(Witness(1), x), (Witness(2), y)]));
+
+    let mut lhs = None;
+    if x_constant {
+        lhs = Some(x);
+    }
+
+    let mut rhs = None;
+    if y_constant {
+        rhs = Some(y);
+    }
+
+    let op = Opcode::BlackBoxFuncCall(blackbox_func_call(lhs, rhs));
+    let opcodes = vec![op];
+    let unconstrained_functions = vec![];
+    let mut acvm =
+        ACVM::new(&StubbedBlackBoxSolver, &opcodes, initial_witness, &unconstrained_functions, &[]);
+    let solver_status = acvm.solve();
+    assert_eq!(solver_status, ACVMStatus::Solved);
+    let witness_map = acvm.finalize();
+
+    witness_map[&Witness(3)]
+}
+
+fn function_input_from_option(
+    witness: Witness,
+    opt_constant: Option<FieldElement>,
+) -> FunctionInput<FieldElement> {
+    opt_constant
+        .map(|constant| FunctionInput::constant(constant, FieldElement::max_num_bits()))
+        .unwrap_or(FunctionInput::witness(witness, FieldElement::max_num_bits()))
+}
+
+fn and_op(x: Option<FieldElement>, y: Option<FieldElement>) -> BlackBoxFuncCall<FieldElement> {
+    let lhs = function_input_from_option(Witness(1), x);
+    let rhs = function_input_from_option(Witness(2), y);
+    BlackBoxFuncCall::AND { lhs, rhs, output: Witness(3) }
+}
+
+fn xor_op(x: Option<FieldElement>, y: Option<FieldElement>) -> BlackBoxFuncCall<FieldElement> {
+    let lhs = function_input_from_option(Witness(1), x);
+    let rhs = function_input_from_option(Witness(2), y);
+    BlackBoxFuncCall::XOR { lhs, rhs, output: Witness(3) }
+}
+
+fn prop_assert_commutative(
```

Comment:
the function (and the others below)  does not assert anything, despite its name.
Having the assert inside the function would reduce overall code size.

Reply:
I attempted to put the assert inside those functions, but ran into some `proptest` type issues.

Using this as an example:

```rust
fn prop_assert_commutative(
    op: impl Fn(Option<FieldElement>, Option<FieldElement>) -> BlackBoxFuncCall<FieldElement>,
    x: (FieldElement, bool),
    y: (FieldElement, bool),
) -> Result<(), proptest::test_runner::TestCaseError> {
    prop_assert_eq!(solve_blackbox_func_call(&op, x, y), solve_blackbox_func_call(&op, y, x))
}
```

This fails, saying that `prop_assert_eq` returns a `()`:

```rust
error[E0308]: mismatched types
   --> acvm-repo/acvm/tests/solver.rs:801:5
    |
801 |     prop_assert_eq!(solve_blackbox_func_call(&op, x, y), solve_blackbox_func_call(&op, y, x))
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Result<(), TestCaseError>`, found `()`
    |
    = note:   expected enum `Result<(), proptest::test_runner::TestCaseError>`
            found unit type `()`
    = note: this error originates in the macro `prop_assert_eq` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0308`.
error: could not compile `acvm` (test "solver") due to previous error
```

But when returning `()`:

```bash
❯ cargo test
   Compiling acvm v0.48.0 (/Users/michaelklein/Coding/rust/noir/acvm-repo/acvm)
error[E0308]: mismatched types
   --> acvm-repo/acvm/tests/solver.rs:801:5
    |
800 | ) -> () {
    |      -- expected `()` because of return type
801 |     prop_assert_eq!(solve_blackbox_func_call(&op, x, y), solve_blackbox_func_call(&op, y, x))
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `()`, found `Result<_, TestCaseError>`
    |
    = note: expected unit type `()`
                    found enum `Result<_, TestCaseError>`
    = note: this error originates in the macro `$crate::prop_assert` which comes from the expansion of the macro `prop_assert_eq` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0308`.
error: could not compile `acvm` (test "solver") due to previous error
```

So it appears that `prop_assert_eq` only behaves as expected when inside a proptest environment where everything is a test. I'm not sure how else to split things up.

Reply:
I don't want to complicate things here, may be it can be refactored in future but duplicating a few asserts is fine for now.